[CALCITE-4917] Add test for 'IS NOT NULL(a) AND a=b' simplification

asolimando opened the pull request:
Simplification (from `RexSimplify` class) is mostly covered in `RexProgramTest`, but tests for expressions of the form `IS NOT NULL(a) AND a=b` (with nullable and not nullable types) seem to be missing.

Since I had to write tests to make sure the simplification was as expected, I assume others might end up doing the same, and that the tests will both act as documentation and it will also protect against regressions.

vlsi commented on core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:
```diff
@@ -1710,6 +1710,17 @@ private void checkExponentialCnf(int n) {
   }
 
+  @Test void testSimplifyAndIsNotNullEquality() {
+    RelDataType intNullableType =
+        typeFactory.createTypeWithNullability(
+            typeFactory.createSqlType(SqlTypeName.INTEGER), true);
+    final RexInputRef i0 = rexBuilder.makeInputRef(intNullableType, 0);
+    final RexInputRef i1 = rexBuilder.makeInputRef(intNullableType, 1);
+
+    checkSimplify2(and(isNotNull(i0), eq(i0, i1)),
```
```suggestion
    checkSimplify2(and(isNotNull(vIntNotNull(0)), eq(vIntNotNull(0), vIntNotNull(1))),
```

vlsi reviewed (approved):
Thanks, it looks good

vlsi commented:
Will merge after https://github.com/apache/calcite/pull/2629

asolimando commented:
@vlsi please hold on to merging this, I'd like to address @julianhyde's comments first, thanks!

chunweilei reviewed (approved):
LGTM

asolimando commented:
@vlsi, @chunweilei: do you want me to squash before merging or will you do it? In case, please use the updated Jira ticket title "Add test for 'a IS NOT NULL AND a = b' simplification" as commit message, the first commit message is stale. Thanks!

julianhyde commented:
@asolimando I acknowledge this is a particular challenge with the current workflow. (@vlsi please don't hit me over the head with it.)
* We want really good commit messages
* We want people to not squash/amend commits, to make them easier to review the deltas. Therefore the contributor can't change the commit message
* Some committers forget to look at JIRA comments before merging (and squashing and rebasing) the PR

asolimando commented:
> * We want people to not squash/amend commits, to make them easier to review the deltas. Therefore the contributor can't change the commit message

I agree on that when the review is on-going, but I have always done my squashing once all the active reviewers mark the PR "approved", just to spare some pain to the reviewers/committers (but also to have a final chance to write an extended commit message if the PR is a bit complex, definitely this is not the case). Does it pose problems too or you were referring to pending reviews?

> * Some committers forget to look at JIRA comments before merging (and squashing and rebasing) the PR

Probably I should have updated the PR title to reflect the new JIRA title.

vlsi commented:
I think we have already discussed "when it is time to squash", and I believe the outcome is like "as soon as one thinks there's a final state, it is a good time to squash".

AFAIK GitHub does show "force push" changes, so "force push" alone does not create problems for the review.

However, there's a notable case: "rebase on top of current master and squash at the same time". That results in "combined diff" where both "changes from master branch" and "changes from the squash" appear in the same diff.

In that case, I prefer to rebase first, force push, then squash and force-push again. It shows as two "force push" events in GitHub UI.

vlsi commented:
> In case, please use the updated Jira ticket title

I think PR author and contributor is in a much better position to update PR/JIRA titles than the reviewer/committer.
All in all, I would like for the contributors to update the PR/JIRA/whatever title so it reflects the actual intention rather than "the first idea which might even be abandoned long ago".

asolimando commented:
I have squashed the commits, updated the commit message to reflect the latest title of the JIRA ticket.
The second force push is a rebase on current master.

julianhyde commented:
@vlsi I agree with that workflow. I have to tell contributors quite often not to amend/rebase while review is ongoing, so we should make an effort to document for contributors the workflow that we would like them to follow.

asolimando commented:
If there are no objections I will merge in the next 24 hours

julianhyde commented:
@asolimando You should announce your intentions on the '[Calcite Release] Preparing for 1.29.0 Release' thread. As RM, Rui is about to lock down master branch.
